### PR TITLE
Add types for current card reducers

### PR DESCRIPTION
**Pull request by oisincoveney**
Fixes types on the dashboard API for public/embed methods, and adds some types to the `dashboard` redux reducers.

**_Why?_**
1. The API should exactly reflect what it needs and what it returns. Right now it's a little too generic. It won't affect actual functionality (because TS) but I don't want anyone to be led the wrong way here.
2. I'm adding types and using the builder pattern so we can use the RTK query endpoints for triggering actions with `builder.addMatcher` for things like `createPublicLink` and `deletePublicLink`. This will be done in a followup PR.

This PR is part 1, which focuses on dashboards. The one for dashboards is [here](https://github.com/metabase/metabase/pull/48399)

### CODE REVIEW

File: frontend/src/metabase-types/api/card.ts
Comment (oisincoveney) [inline frontend/src/metabase-types/api/card.ts:1]: I changed these according to the actual return types I saw on the network tab when creating and deleting public links, and viewing embeddable cards.

File: frontend/src/metabase/query_builder/actions/sharing.ts [hunk 1]
```diff
@@ -29,19 +27,23 @@ export const deletePublicLink = createThunkAction(
...
-    }),
-);
-
-export const UPDATE_EMBEDDING_PARAMS = "metabase/card/UPDATE_EMBEDDING_PARAMS";
-export const updateEmbeddingParams = createAction(
-  UPDATE_EMBEDDING_PARAMS,
-  ({ id }: CardIdPayload, embedding_params: EmbedOptions) =>
-    CardApi.update({ id, embedding_params }),
+export const updateEnableEmbedding = createAsyncThunk<
+  Pick<Card, "id" | "enable_embedding"> & { uuid: Card["public_uuid"] },
```
Comment (oisincoveney) [inline frontend/src/metabase/query_builder/actions/sharing.ts:31]: We can probably improve this one. Should we just make a new type?

File: frontend/src/metabase/query_builder/reducers-typed.ts
Comment (oisincoveney) [inline frontend/src/metabase/query_builder/reducers-typed.ts:1]: I just took the reducer from `reducers.js` and converted it to a builder pattern. The one thing that's annoying here is that the initial state is `null`, and I don't particularly want to change that and find out the consequences of that. That's why I have a bunch of if statements. Maybe we can move that into a function but it feels weird to something like that for this case

File: frontend/src/metabase/query_builder/actions/sharing.ts (hunk 1 above)
Comment (heypoom) [reply frontend/src/metabase/query_builder/actions/sharing.ts:31]: Making a new type sounds good

**Review by heypoom — APPROVED** [review]
Looks good overall. I'm a bit unsure about the use of `EmbeddingParameters` in place of `EmbedOptions` though

File: frontend/src/metabase/query_builder/actions/sharing.ts [hunk 2]
```diff
@@ -29,19 +27,23 @@ export const deletePublicLink = createThunkAction(
...
+    id,
+    enable_embedding,
+  });
+
+  return { id, uuid: response.uuid, enable_embedding };
+});
+
+export const updateEmbeddingParams = createAsyncThunk<
+  Card,
+  { id: CardId; embedding_params: EmbeddingParameters }
```
[the remark this answers is not in the document]
Comment (oisincoveney) [reply frontend/src/metabase/query_builder/actions/sharing.ts:44]: Let me just double check that. I looked at the payloads and what I saw was that we were passing the parameters, not the display options. So I'm fairly sure this is what we should do

File: frontend/src/metabase/query_builder/actions/sharing.ts [hunk 3]
```diff
@@ -29,19 +27,23 @@ export const deletePublicLink = createThunkAction(
...
+export const updateEnableEmbedding = createAsyncThunk<
+  Pick<Card, "id" | "enable_embedding"> & { uuid: Card["public_uuid"] },
+  Pick<Card, "id" | "enable_embedding">
+>("metabase/card/UPDATE_ENABLE_EMBEDDING", async ({ id, enable_embedding }) => {
+  const response = await CardApi.update({
+    id,
+    enable_embedding,
+  });
+
+  return { id, uuid: response.uuid, enable_embedding };
```
[the remark this answers is not in the document]
Comment (oisincoveney) [reply frontend/src/metabase/query_builder/actions/sharing.ts:39]: Yep this is what the reducers expect. I'm going to check the return value from the API again just to make sure it's not returning the whole question object (I checked yesterday but I'm confusing the dashboard and card APIs at this point).

File: frontend/src/metabase/query_builder/actions/sharing.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (oisincoveney) [reply frontend/src/metabase/query_builder/actions/sharing.ts:44]: Ok, I created a native question and published it. There's two calls when published, the `enable_embedding` and `embedding_params` calls. The `embedding_params` call sends the actual parameters, not the embed options, so I think we should be ok.

File: frontend/src/metabase/query_builder/actions/sharing.ts (hunk 1 above)
Comment (oisincoveney) [reply frontend/src/metabase/query_builder/actions/sharing.ts:31]: actually.. I'll leave this. I'll make the type when I migrate the embedding endpoints to RTK in [this PR](https://github.com/metabase/metabase/pull/48391)

**Pull request merged**